File: app/src/main/java/com/mkulesh/onpc/config/Configuration.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2022 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.config;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.mkulesh.onpc.iscp.BroadcastSearch;
import com.mkulesh.onpc.iscp.State;
import com.mkulesh.onpc.iscp.messages.InputSelectorMsg;
import com.mkulesh.onpc.iscp.messages.NetworkServiceMsg;
import com.mkulesh.onpc.iscp.messages.ReceiverInformationMsg;
import com.mkulesh.onpc.iscp.messages.ServiceType;
import com.mkulesh.onpc.utils.Logging;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;

import static com.mkulesh.onpc.utils.Utils.getStringPref;

public class Configuration
{
    public static final boolean ENABLE_MOCKUP = false;

    static final String PROTO_TYPE = "proto_type";
    private static final String SERVER_NAME = "server_name";
    private static final String SERVER_PORT = "server_port";

    static final String MODEL = "model";
    private static final String ACTIVE_ZONE = "active_zone";

    static final String DEVICE_SELECTORS = "device_selectors";
    private static final String SELECTED_DEVICE_SELECTORS = "selected_device_selectors";

    private static final String AUTO_POWER = "auto_power";
    static final String FRIENDLY_NAMES = "pref_friendly_names";
    static final String NETWORK_SERVICES = "network_services";
    private static final String SELECTED_NETWORK_SERVICES = "selected_network_services";

    private static final String KEEP_SCREEN_ON = "keep_screen_on";
    private static final String BACK_AS_RETURN = "back_as_return";
    private static final String ADVANCED_QUEUE = "advanced_queue";
    private static final String KEEP_PLAYBACK_MODE = "keep_playback_mode";
    private static final String EXIT_CONFIRM = "exit_confirm";
    private static final String DEVELOPER_MODE = "developer_mode";

    private final SharedPreferences preferences;

    private String deviceName;
    private int devicePort;

    public final CfgAppSettings appSettings = new CfgAppSettings();
    public final CfgAudioControl audioControl = new CfgAudioControl();
    public final CfgFavoriteConnections favoriteConnections = new CfgFavoriteConnections();
    public final CfgFavoriteShortcuts favoriteShortcuts = new CfgFavoriteShortcuts();

    public Configuration(Context context)
    {
        preferences = PreferenceManager.getDefaultSharedPreferences(context);

        deviceName = preferences.getString(Configuration.SERVER_NAME, "");
        devicePort = preferences.getInt(Configuration.SERVER_PORT, BroadcastSearch.ISCP_PORT);

        appSettings.setPreferences(preferences);

        audioControl.setPreferences(preferences);
        audioControl.read(context);

        favoriteConnections.setPreferences(preferences);
        favoriteConnections.read();

        favoriteShortcuts.setPreferences(preferences);
        favoriteShortcuts.read();
    }

    public String getDeviceName()
    {
        return deviceName;
    }

    public int getDevicePort()
    {
        return devicePort;
    }

    @SuppressLint("SetTextI18n")
    public String getDevicePortAsString()
    {
        return Integer.toString(devicePort);
    }

    public void saveDevice(final String device, final int port)
    {
        deviceName = device;
        devicePort = port;
        SharedPreferences.Editor prefEditor = preferences.edit();
        prefEditor.putString(SERVER_NAME, device);
        prefEditor.putInt(SERVER_PORT, port);
        prefEditor.apply();
    }

    public void initActiveZone(int defaultActiveZone)
    {
        final String activeZone = getStringPref(preferences, ACTIVE_ZONE, "");
        if (activeZone.isEmpty())
        {
            setActiveZone(defaultActiveZone);
        }
    }

    @SuppressLint("ApplySharedPref")
    public void setActiveZone(int zone)
    {
        SharedPreferences.Editor prefEditor = preferences.edit();
        prefEditor.putString(ACTIVE_ZONE, Integer.toString(zone));
        prefEditor.commit();
    }

    public int getZone()
    {
        try
        {
            final String activeZone = getStringPref(preferences, ACTIVE_ZONE, "");
            return (activeZone.isEmpty()) ? ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE : Integer.parseInt(activeZone);
        }
        catch (Exception e)
        {
            return ReceiverInformationMsg.DEFAULT_ACTIVE_ZONE;
        }
    }

    public void setReceiverInformation(@NonNull State state)
    {
        SharedPreferences.Editor prefEditor = preferences.edit();

        Logging.info(this, "Save receiver information");
        Logging.info(this, "    Network protocol: " + state.protoType.name());
        prefEditor.putString(PROTO_TYPE, state.protoType.name());
        final String model = state.getModel();
        Logging.info(this, "    Model: " + model);
        if (!model.isEmpty())
        {
            prefEditor.putString(MODEL, model);
        }
        if (!state.networkServices.isEmpty())
        {
            final StringBuilder str = new StringBuilder();
            for (final String p : state.networkServices.keySet())
            {
                if (!str.toString().isEmpty())
                {
                    str.append(",");
                }
                str.append(p);
            }
            Logging.info(this, "    Network services: " + str);
            prefEditor.putString(NETWORK_SERVICES, str.toString());
        }
        List<ReceiverInformationMsg.Selector> deviceSelectors = state.cloneDeviceSelectors();
        if (!deviceSelectors.isEmpty())
        {
            final StringBuilder str = new StringBuilder();
            for (ReceiverInformationMsg.Selector d : deviceSelectors)
            {
                if (!str.toString().isEmpty())
                {
                    str.append(",");
                }
                str.append(d.getId());
                prefEditor.putString(DEVICE_SELECTORS + "_" + d.getId(), d.getName());
            }
            Logging.info(this, "    Device selectors: " + str);
            prefEditor.putString(DEVICE_SELECTORS, str.toString());
        }

        prefEditor.apply();

        favoriteConnections.updateIdentifier(state);
    }

    @NonNull
    static String getSelectedDeviceSelectorsParameter(final SharedPreferences p)
    {
        return Configuration.SELECTED_DEVICE_SELECTORS + "_" + p.getString(MODEL, "NONE");
    }

    @NonNull
    public ArrayList<ReceiverInformationMsg.Selector> getSortedDeviceSelectors(
            boolean allItems,
            @NonNull InputSelectorMsg.InputType activeItem,
            @NonNull final List<ReceiverInformationMsg.Selector> defaultItems)
    {
        final ArrayList<ReceiverInformationMsg.Selector> result = new ArrayList<>();
        final ArrayList<String> defItems = new ArrayList<>();
        for (ReceiverInformationMsg.Selector i : defaultItems)
        {
            defItems.add(i.getId());
        }
        final String par = getSelectedDeviceSelectorsParameter(preferences);
        for (CheckableItem sp : CheckableItem.readFromPreference(preferences, par, defItems))
        {
            final boolean visible = allItems || sp.checked ||
                    (activeItem != InputSelectorMsg.InputType.NONE && activeItem.getCode().equals(sp.code));
            for (ReceiverInformationMsg.Selector i : defaultItems)
            {
                if (visible && i.getId().equals(sp.code))
                {
                    result.add(i);
                }
            }
        }
        return result;
    }

    public boolean isAutoPower()
    {
        return preferences.getBoolean(AUTO_POWER, false);
    }

    public boolean isFriendlyNames()
    {
        return preferences.getBoolean(FRIENDLY_NAMES, true);
    }

    @NonNull
    static String getSelectedNetworkServicesParameter(final SharedPreferences p)
    {
        return Configuration.SELECTED_NETWORK_SERVICES + "_" + p.getString(MODEL, "NONE");
    }

    @NonNull
    public ArrayList<NetworkServiceMsg> getSortedNetworkServices(
            @NonNull ServiceType activeItem,
            @NonNull final List<NetworkServiceMsg> defaultItems)
    {
        final ArrayList<NetworkServiceMsg> result = new ArrayList<>();
        final ArrayList<String> defItems = new ArrayList<>();
        for (NetworkServiceMsg i : defaultItems)
        {
            defItems.add(i.getService().getCode());
        }
        final String par = getSelectedNetworkServicesParameter(preferences);
        for (CheckableItem sp : CheckableItem.readFromPreference(preferences, par, defItems))
        {
            final boolean visible = sp.checked ||
                    (activeItem != ServiceType.UNKNOWN && activeItem.getCode().equals(sp.code));
            for (NetworkServiceMsg i : defaultItems)
            {
                if (visible && i.getService().getCode().equals(sp.code))
                {
                    result.add(new NetworkServiceMsg(i));
                }
            }
        }
        return result;
    }

    public boolean isKeepScreenOn()
    {
        return preferences.getBoolean(Configuration.KEEP_SCREEN_ON, false);
    }

    public boolean isBackAsReturn()
    {
        return preferences.getBoolean(Configuration.BACK_AS_RETURN, false);
    }

    public boolean isAdvancedQueue()
    {
        return preferences.getBoolean(Configuration.ADVANCED_QUEUE, false);
    }

    public boolean keepPlaybackMode()
    {
        return preferences.getBoolean(Configuration.KEEP_PLAYBACK_MODE, false);
    }

    public boolean isExitConfirm()
    {
        return preferences.getBoolean(Configuration.EXIT_CONFIRM, false);
    }

    public boolean isDeveloperMode()
    {
        return preferences.getBoolean(DEVELOPER_MODE, false);
    }
}

File: lib/config/Configuration.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2024 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "package:shared_preferences/shared_preferences.dart";

import "../constants/Version.dart";
import "../iscp/ConnectionIf.dart";
import "../iscp/StateManager.dart";
import "../iscp/messages/EnumParameterMsg.dart";
import "../iscp/messages/InputSelectorMsg.dart";
import "../iscp/state/ReceiverInformation.dart";
import "../utils/Convert.dart";
import "../utils/Logging.dart";
import "../utils/Pair.dart";
import "../utils/Platform.dart";
import "CfgAppSettings.dart";
import "CfgAudioControl.dart";
import "CfgFavoriteConnections.dart";
import "CfgFavoriteShortcuts.dart";
import "CfgModule.dart";
import "CfgRiCommands.dart";

class Configuration extends CfgModule
{
    static const String CONFIGURATION_EVENT = "CONFIG";
    String appVersion = Version.NAME;

    // Connection options
    static const Pair<String, String> SERVER_NAME = Pair<String, String>("server_name", "");
    String _deviceName = SERVER_NAME.item2;

    String get getDeviceName
    => _deviceName;

    static const Pair<String, int> SERVER_PORT = Pair<String, int>("server_port", 60128);
    int _devicePort = SERVER_PORT.item2;

    int get getDevicePort
    => _devicePort;

    bool get isDeviceValid
    => _deviceName.isNotEmpty && _devicePort > 0;

    static const Pair<String, int> ACTIVE_ZONE = Pair<String, int>("active_zone", 0);
    int _activeZone = ACTIVE_ZONE.item2;

    int get activeZone
    => _activeZone;

    set activeZone(int value)
    {
        _activeZone = value;
        saveIntegerParameter(ACTIVE_ZONE, value);
    }

    // Device options
    static const Pair<String, bool> AUTO_POWER = Pair<String, bool>("auto_power", false);
    bool _autoPower = AUTO_POWER.item2;

    bool get autoPower
    => _autoPower;

    static const Pair<String, bool> FRIENDLY_NAMES = Pair<String, bool>("friendly_names", true);
    bool _friendlyNames = FRIENDLY_NAMES.item2;

    bool get friendlyNames
    => _friendlyNames;

    static const Pair<String, String> PROTO_TYPE = Pair<String, String>("proto_type", "ISCP");
    static const Pair<String, String> MODEL = Pair<String, String>("model", "NONE");
    static const Pair<String, String> DEVICE_FRIENDLY_NAME = Pair<String, String>("device_friendly_name", "");

    static const String NETWORK_SERVICES = "network_services";
    static const String SELECTED_NETWORK_SERVICES = "selected_network_services";

    static const String DEVICE_SELECTORS = "device_selectors";
    static const String SELECTED_DEVICE_SELECTORS = "selected_device_selectors";
    static const String MANUAL_DEVICE_SELECTORS = "manual_device_selectors";

    // Advanced options
    static const Pair<String, bool> KEEP_SCREEN_ON = Pair<String, bool>("keep_screen_on", false); // For Android only
    bool _keepScreenOn = KEEP_SCREEN_ON.item2;

    bool get keepScreenOn
    => _keepScreenOn;

    static const Pair<String, bool> SHOW_WHEN_LOCKED = Pair<String, bool>("show_when_locked", false); // For Android only
    bool _showWhenLocked = SHOW_WHEN_LOCKED.item2;

    bool get showWhenLocked
    => _showWhenLocked;

    static const Pair<String, bool> BACK_AS_RETURN = Pair<String, bool>("back_as_return", true); // For Android only
    bool _backAsReturn = BACK_AS_RETURN.item2;

    bool get backAsReturn
    => _backAsReturn;

    static const Pair<String, bool> ADVANCED_QUEUE = Pair<String, bool>("advanced_queue", true);
    bool _advancedQueue = ADVANCED_QUEUE.item2;

    bool get isAdvancedQueue
    => _advancedQueue;

    static const Pair<String, bool> KEEP_PLAYBACK_MODE = Pair<String, bool>("keep_playback_mode", false);
    bool _keepPlaybackMode = KEEP_PLAYBACK_MODE.item2;

    bool get keepPlaybackMode
    => _keepPlaybackMode;

    static const Pair<String, bool> EXIT_CONFIRM = Pair<String, bool>("exit_confirm", false);
    bool _exitConfirm = EXIT_CONFIRM.item2;

    bool get exitConfirm
    => _exitConfirm;

    static const Pair<String, bool> DEVELOPER_MODE = Pair<String, bool>("developer_mode", false);
    bool _developerMode = DEVELOPER_MODE.item2;

    bool get developerMode
    => _developerMode;

    // configuration modules
    late CfgAppSettings appSettings;
    late CfgAudioControl audioControl;
    late CfgFavoriteConnections favoriteConnections;
    late CfgFavoriteShortcuts favoriteShortcuts;
    late CfgRiCommands riCommands;

    Configuration(final SharedPreferences preferences) : super(preferences)
    {
        appVersion = Version.NAME;
        appSettings = CfgAppSettings(preferences);
        audioControl = CfgAudioControl(preferences);
        favoriteConnections = CfgFavoriteConnections(preferences);
        favoriteShortcuts = CfgFavoriteShortcuts(preferences);
        riCommands = CfgRiCommands(preferences);
        Logging.info(this, "Application started: " + appVersion + ", OS: " + Platform.operatingSystem);
    }

    @override
    void read({ProtoType? protoType})
    {
        Logging.info(this, "Reading configuration...");

        // Connection options
        _deviceName = getString(SERVER_NAME, doLog: true);
        _devicePort = getInt(SERVER_PORT, doLog: true);
        _activeZone = getInt(ACTIVE_ZONE, doLog: true);

        // Device options
        _autoPower = getBool(AUTO_POWER, doLog: true);
        _friendlyNames = getBool(FRIENDLY_NAMES, doLog: true);

        // Advanced options
        _keepScreenOn = Platform.isAndroid ? getBool(KEEP_SCREEN_ON, doLog: true) : false;
        _showWhenLocked = Platform.isAndroid ? getBool(SHOW_WHEN_LOCKED, doLog: true) : false;
        _backAsReturn = Platform.isAndroid ? getBool(BACK_AS_RETURN, doLog: true) : false;
        _advancedQueue = getBool(ADVANCED_QUEUE, doLog: true);
        _keepPlaybackMode = getBool(KEEP_PLAYBACK_MODE, doLog: true);
        _exitConfirm = Platform.isAndroid ? getBool(EXIT_CONFIRM, doLog: true) : false;
        _developerMode = getBool(DEVELOPER_MODE, doLog: true);

        // configuration modules
        appSettings.read(protoType: protoType);
        audioControl.read();
        favoriteConnections.read();
        favoriteShortcuts.read();
        riCommands.read();
    }

    void readHomeWidgetCfg()
    {
        Logging.info(this, "Reading home widget configuration...");

        // Connection options
        _deviceName = getString(SERVER_NAME, doLog: true);
        _devicePort = getInt(SERVER_PORT, doLog: true);
        _activeZone = getInt(ACTIVE_ZONE, doLog: true);
        audioControl.read();
    }

    void saveDevice(final String device, final int port) async
    {
        _deviceName = device;
        _devicePort = port;
        await preferences.setString(SERVER_NAME.item1, device);
        await preferences.setInt(SERVER_PORT.item1, port);
    }

    @override
    void setReceiverInformation(StateManager stateManager)
    {
        final ReceiverInformation state = stateManager.state.receiverInformation;
        String m = state.model;
        if (m.isEmpty)
        {
            m = MODEL.item2;
        }
        Logging.info(this, "Save receiver information");
        saveStringParameter(PROTO_TYPE, Convert.enumToString(stateManager.protoType), prefix: "  ");
        saveStringParameter(MODEL, m, prefix: "  ");
        saveStringParameter(DEVICE_FRIENDLY_NAME, state.getDeviceName(true), prefix: "  ");
        if (state.networkServices.isNotEmpty)
        {
            String str = "";
            state.networkServices.forEach((p)
            {
                if (str.isNotEmpty)
                {
                    str += ",";
                }
                str += p.getId;
            });
            saveStringParameter(Pair<String, String>(NETWORK_SERVICES, ""), str, prefix: "  ");
        }
        if (state.deviceSelectors.isNotEmpty)
        {
            String str = "";
            state.deviceSelectors.forEach((d)
            {
                if (str.isNotEmpty)
                {
                    str += ",";
                }
                str += d.getId;
                preferences.setString(DEVICE_SELECTORS + "_" + d.getId, d.getName);
            });
            saveStringParameter(Pair<String, String>(DEVICE_SELECTORS, ""), str, prefix: "  ");
        }
        appSettings.setReceiverInformation(stateManager);
        audioControl.setReceiverInformation(stateManager);
        favoriteConnections.setReceiverInformation(stateManager);
        favoriteShortcuts.setReceiverInformation(stateManager);
    }

    void saveManualDeviceSelector(final EnumItem<InputSelector> item, final String name)
    {
        final Pair<String, String> par = Pair<String, String>(MANUAL_DEVICE_SELECTORS + "_" + item.getCode, "");
        name.isNotEmpty ? saveStringParameter(par, name) : deleteParameter(par);
    }

    String deviceSelectorName(final EnumItem<InputSelector> item, {bool useFriendlyName = false, String friendlyName = ""})
    {
        final Pair<String, String> par = Pair<String, String>(MANUAL_DEVICE_SELECTORS + "_" + item.getCode, "");
        final String manName = getString(par);
        if (manName.isNotEmpty)
        {
            return manName;
        }
        final String defName = item.description.toUpperCase();
        if (useFriendlyName)
        {
            return friendlyName.isNotEmpty ? friendlyName :
                getStringDef(Configuration.DEVICE_SELECTORS + "_" + item.getCode, defName);
        }
        return defName;
    }

    void saveDeviceFriendlyName(final ReceiverInformation state)
    {
        saveStringParameter(DEVICE_FRIENDLY_NAME, state.getDeviceName(true));
    }

    ProtoType get protoType
    => Convert.stringToProtoType(getString(PROTO_TYPE));
}
